Add a8c /callout block

aaronyan opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Related: p7H4VZ-3qe-p2

## Proposed Changes

Adds a new callout block (`a8c/callout`) to the O2 blocks collection for creating styled callout boxes with icon, header, and content sections.

### Features
- **Pre-built Templates**: 5 built-in templates (Info, Tip, Example, Warning, Custom) with predefined colors and icons
- **Customizable Styling**: Custom color picker and icon selector for the Custom template
- **Consistent Design**: Uses Automattic color palette and Dashicons for visual consistency

<img width="1276" alt="Screenshot" src="https://github.com/user-attachments/assets/acbcdea4-5d42-4c2e-b4cb-c546bd03bcf3" />

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

This block enhances content authoring by providing a standardized way to create attention-grabbing callout sections for documentation, tutorials, and informational content on P2s.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### Basic Block Creation
1. Open the block editor and click the "+" button to add a new block
2. Search for "Callout" in the block inserter
3. Verify the block appears in the "a8c" category with a megaphone icon
4. Insert the block and confirm it creates a callout with:
   - An icon (megaphone by default)
   - A "Header" placeholder
   - An "Enter details here…" placeholder

#### Template Functionality
1. Select the callout block and open the block settings sidebar
2. In the "Templates" panel, test switching between templates:
   - **Info**: Blue color with info icon
   - **Tip**: Celadon color with lightbulb icon
   - **Example**: Purple color with edit-page icon
   - **Warning**: Red color with flag icon
   - **Custom**: Gray color with megaphone icon
3. Verify the icon and color change immediately when switching templates
4. Confirm that switching from Custom to another template clears any custom settings

#### Custom Template Features (Custom template only)
1. Select the Custom template
2. Test the **Color** panel:
   - Click different colors from the palette
   - Verify the callout border and icon color update
   - Test custom color input by entering a hex value
3. Test the **Icon** panel:
   - Click different icons in the grid
   - Verify the icon updates in the callout
   - Confirm selected icon shows a colored border
   - Check that the "Selected:" text shows the current icon name

#### Content Editing
1. Click in the header area and type a custom header (type a short and long header)
2. Click in the content area and add various block types:
   - Paragraphs
   - Lists
   - Images
   - Other blocks
3. Verify all content renders properly within the callout styling
   - The header text wraps to multiple lines when it exceeds the available width
   - The icon remains vertically centered relative to the first line of the header text
   - The spacing between the icon and header text remains consistent regardless of header length
5. Confirm the icon section cannot be deleted or moved

#### Prefix Transform
1. Create a new paragraph block
2. Type `/callout` and press Enter
3. Verify it transforms into a callout block
4. Test the same with `/callout` in an existing paragraph

#### Visual Verification
1. Check that callouts display properly on both:
   - Editor view
   - Published/preview view
2. Verify responsive behavior on different screen sizes
3. Confirm proper spacing and typography
4. Test with different WordPress themes

#### Persistence
1. Create a callout with custom settings
2. Save the post and reload the editor
3. Verify all settings (template, color, icon, content) are preserved
4. Test undo/redo functionality

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aaronyan committed:
Add callout block to o2-blocks

aaronyan committed:
Refactor callout block into dependent header and icon components

Split monolithic callout block to prevent users from accidentally
removing required CSS classes by making header and icon non-editable
child components. Fix dashicon linting by using CSS classes instead
of Dashicon component.

aaronyan committed:
Fix header to wrap correctly for long text

dmsnell commented on apps/o2-blocks/src/callout-header/view.js:
```diff
@@ -0,0 +1,5 @@
+ * Callout Header Block - Frontend Script
+ *
+ * No frontend JavaScript needed for this block.
+ */
```
we probably don’t want or need to create empty files that aren’t necessary.

dmsnell reviewed (commented):
I’m no expert here so it would be worth asking someone who knows, but I think that using `p` for the heading would be less desirable than using an element which assistive technologies will identify as a heading. for example, I think something like this would work better, but again I know very little about this and only want to poke at it and see if you can just double-check on this without too much effort.

```html
<aside>
<h2>...</h2>
<p>...</p>
</aside>
```

there’s some potential for something like a `aria-labelledby` attribute on the `aside` to associate it with the header, but it’s always hard to know if the use of the `aria-` attributes helps or hurts.

---

the use of the template here is a bit curious. in a relatively simple block like this with relatively static elements, there’s only a real need for a single block type unless I’m overlooking something.

whereas you have separate blocks which each bring their own structure, I would expect it possible to embed those into into the single `a8c/callout` block without creating three. this can be done directly in the `Edit` function or parts of those pieces can be created as React components and then included via normal React patterns.

this would eliminate the need for `InnerBlocks` unless you wanted to support what I originally proposed, which was allowing people to add more than a paragraph (perhaps someone wants a list in a callout).

besides that, the extreme level of modularization and templating is something I find a bit confusing and verbose. perhaps your LLM tooling can reign in its suggestions and considerate into something simpler.

---

I’m short on time at the moment (on a meetup) so I wanted to leave some comments but can’t give this the proper attention it warrants. Hope this helps and gets you at least a few steps forward.

aaronyan committed:
Change to aside and h2 elements and support aria-labelledby attribute

aaronyan committed:
Reduced to single block and removed aria-labelledby attribute

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,331 @@
+	transforms,
+	edit: Edit,
+	save: Save,
+} );
```
it would be helpful to create a `blocks.json` file for this so that if we need information about it from PHP then we can read it.

aaronyan replied:
Sounds good! Added in https://github.com/Automattic/wp-calypso/pull/104352/commits/bdbc2aea9c12e310e70a18881464ee39e2063b6c!

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,331 @@
+			setAttributes( { headerContent: newHeaderContent } );
+		},
+		[ setAttributes ]
+	);
```
for as simple as all of these are it would be helpful to avoid creating the callbacks. they can be inlined when setting the handlers.

also it should be fine to call `setAttributes()` even when the value isn’t changed, as the editor can already handle non-update updates.

aaronyan replied:
Makes sense! Making these handlers inline given how simple they are!

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,331 @@
+
+const getCurrentColor = ( templateName, customColor ) => {
+	return templateName === 'custom'
+		? normalizeColor( customColor ) || CALLOUT_TEMPLATE_DEFAULT_COLOR
```
try examining the data flow through the functions that were generated here and consider the `||` defaulting value. there’s something suspect in what’s written.

aaronyan replied:
Doh, thanks for the catch. Resolved in https://github.com/Automattic/wp-calypso/pull/104352/commits/bdbc2aea9c12e310e70a18881464ee39e2063b6c

dmsnell reviewed (commented):
I left a few comments that I hope provide some path forward, while being limited. I wrote them with the expectation that you might refresh this with whatever LLM-based tools you are using to generate the code.

Try iterating on the code incorporating the comments and then feel free to ask the LLM tools to provide additional review (but of course, please consider what they write and ask if it’s good advice or not).

Happy to look at this again after iterating on it.

aaronyan committed:
Create blocks.json, inline handlers, fix defaults

aaronyan committed:
Add background color option

aaronyan committed:
Memoize custom styles and icon picker and add aria labels to icons

aaronyan commented:
Thanks so much for the additional review, @dmsnell! I really appreciate the advice and helping me walkthrough this experience. I'm learning a ton!

Summary of the changes:
- Created blocks.json, moved handlers to be inline, and fixed default values
- Added background color picker
- Added some optimizations based on recommendations from the LLM (memoized custom styles and icon grid) 
- Added aria labels for icon buttons

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,349 @@
+						aria-label={ `Select ${ iconName.replace( 'dashicons-', '' ) } icon${
+							isSelected ? ' (currently selected)' : ''
+						}` }
+						title={ iconName }
```
what is the reason for removing the `dashicons-` prefix on the `aria-label` but then leaving it in for the `title`? why did you make this choice?

dmsnell commented on apps/o2-blocks/src/callout/constants.js:
```diff
@@ -0,0 +1,123 @@
+	{ name: __( 'Celadon' ), color: COLORS[ 'Celadon 40' ] },
+	{ name: __( 'Gray 40' ), color: COLORS[ 'Gray 40' ] },
+	{ name: __( 'Gray 60' ), color: COLORS[ 'Gray 60' ] },
+	{ name: __( 'Gray 80' ), color: COLORS[ 'Gray 80' ] },
```
what is the shape of the data in the colors package? does it not have the name of the colors? it seems like we’re recreating a lot of the data from that, though I’m definitely not familiar with it so I don’t know. is it missing color names that it should have?

dmsnell commented on apps/o2-blocks/src/callout/constants.js:
```diff
@@ -0,0 +1,123 @@
+import { __ } from '@wordpress/i18n';
+
+// Extract colors from Color Studio
+const COLORS = colorStudio.colors;
```
what is the purpose of extracting this as a VARIABLE apart from the import?

dmsnell commented on apps/o2-blocks/src/callout/block.json:
```diff
@@ -0,0 +1,45 @@
+		},
+		"customColor": {
+			"type": "string",
+			"default": "#50575e"
```
why are these custom attributes set with default values? how do we indicate that there’s no custom color?

dmsnell commented on apps/o2-blocks/src/callout/constants.js:
```diff
@@ -0,0 +1,123 @@
+	'dashicons-visibility',
+	'dashicons-hidden',
+
+	// Work related icons
```
How did you derive these categories? Am I missing something about a bar chart being work related or that coffee is about social media?

dmsnell commented on apps/o2-blocks/src/callout/constants.js:
```diff
@@ -0,0 +1,123 @@
+	'dashicons-format-audio',
+	'dashicons-camera-alt',
+	'dashicons-video-alt2',
+];
```
did you examine if it’s feasible to let people set their own Emoji instead of choosing from this hand-picked set of dashicons? did you look into whether we can load our custom emoji loaded from the Slack list?

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,349 @@
+
+const getIcon = ( templateName, customIcon = '' ) => {
+	if ( customIcon ) {
+		return customIcon.startsWith( 'dashicons-' ) ? customIcon : `dashicons-${ customIcon }`;
```
what role do dashicons play in this? it seems strange to me that we would be passing around custom icons that may or may not have a `dashicons-` prefix.

dmsnell commented on apps/o2-blocks/src/callout/editor.js:
```diff
@@ -0,0 +1,349 @@
+	}
+
+	return /^#[0-9A-F]{6}$/i.test( color );
+};
```
“valid” is a loaded term I usually like to be careful about. in this case, this function is not determining whether a given input is a valid color, but actually whether it’s a valid RGB hex triplet.

more broadly though, this code is permeated with fallback handlers for failure cases that we really don’t expect, right? the block itself should never create invalid attribute updates — it’s the role of the `Edit()` function to produce valid, atomic, updates. If `Edit()` produces invalid attributes then it’s a bug in the code, not an expected runtime condition.

The one exception is when loading the block from the stored attributes from the `post_content`, where updates might have happened from outside of the block code, but still, those should be handled in any block invalidation or block transform upon load. At worst, the block remains frozen in the invalid state. At best, it silently loads a usable value.

What are your thoughts here?

dmsnell reviewed (commented):
There’s a lot of abstraction happening within the `constants` file. I wonder what your thoughts are on the way this block pushes around the data it works with. Do you like the way it flows? Did you have other ideas for how to communicate the various settings? Do you like the way there are many split-up React components threading data _through_ them (also see the term, “prop-drilling”).

Let me know what you think about this.
